### ISSUE TITLE
Rule: No Object Manager in classes, except for factories, proxies, builders

### ISSUE THREAD

**Issue by jissereitsma** [extdn/extdn-phpcs#8]
The Object Manager should not be injected directly through constructor DI into a class, except for when that classes appears to be a factory, proxy, builder or similar pattern.

**Comment by renttek** [comment]
What about the use of ObejectManager::getInstance and the new keyword ?

**Comment by jissereitsma** [comment]
The call `ObjectManager::getInstance` gives you an instance of the Object Manager, which is already mentioned by the rule topic. So, yes. As for the `new` keyword, this is different discussion and would need a new issue instead.

**Comment by ihor-sviziev** [comment]
Why for factories, proxies, builders? You mean just inside those classes object manager can be used? Or for creation factory we can use object manager? If second option - I'm totally disagree

**Comment by jissereitsma** [comment]
Yes, I meant the first. Whenever, you create a factory, proxy or builder class, there is an injection of the object manager in its constructor. For reference, see the source code of Magento its core factories and proxies (generated or not). The same applies to builders. So the design pattern more or less makes it legal to inject the object manager. For other classes, there is no such pattern.

I'm not sure what you mean with *creation factory*. Do you have a code sample for this?

**Comment by larsroettig** [comment]
@schmengler i will work on this

**Comment by fooman** [comment]
I guess ObjectManager usage should also be allowed in tests.

**Comment by fooman** [comment]
But maybe it's a broader discussion if the ruleset should even run over tests at all (maybe it just falls back to PSR-2 for example).

**Comment by navarr** [comment]
Magento themselves care about coding standards in tests - at least for PHPCS/PHPMD - so this probably should too

**Comment by lenaorobei** [comment]
Maybe it'll be helpful https://github.com/magento/marketplace-eqp/blob/master/MEQP2/Sniffs/Classes/ObjectManagerSniff.php

There are no exceptions in this sniff, but they can be easily added.

**Comment by jissereitsma** [comment]
@lenaorobei The EQP2 rule mentioned could be included as well, but simply checks whether a variable with the name `objectmanager` (or something) is used in the core.

To all, I have now created a PR for a rule that checks the constructor instead: https://github.com/extdn/extdn-phpcs/pull/48

**Comment by jissereitsma** [comment]
I have added a label for debate, because I have added some additional utilities (`Utils/`) in this PR that use the Reflection API to allow for the class constructor to be inspected in a more flexible way. I'm not sure if that's the path to go for.
